### ISSUE TITLE
Exception from destructor in CREATE DICTIONARY

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#12158]
```
2020.07.06 06:04:00.515512 [ 268 ] {} <Fatal> BaseDaemon: (version 20.6.1.3972, build id: 7FD797363ABE3AB3) (from thread 121973) Terminate called for uncaught exception:
Poco::Exception. Code: 1000, e.code() = 2, e.displayText() = File not found: /var/lib/clickhouse/metadata/dictdb/restricted_dict.sql.tmp, Stack trace (when copying this message, always include the lines below):

0. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/exception:0: std::exception::capture() @ 0x15660625 in /usr/bin/clickhouse
1. /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Exception.cpp:28: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x2196aac1 in /usr/bin/clickhouse
2. /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Exception.cpp:164: Poco::FileNotFoundException::FileNotFoundException(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x21977396 in /usr/bin/cli
3. /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/File_UNIX.cpp:499: Poco::FileImpl::handleLastErrorIm
2020.07.06 06:04:00.515583 [ 268 ] {} <Trace> BaseDaemon: Received signal 6
2020.07.06 06:04:00.515661 [ 634143 ] {} <Fatal> BaseDaemon: ########################################
2020.07.06 06:04:00.515713 [ 634143 ] {} <Fatal> BaseDaemon: (version 20.6.1.3972, build id: 7FD797363ABE3AB3) (from thread 121973) (query_id: 03388d9b-0674-46f7-92c7-24bce88a4357) Received signal Aborted (6)
2020.07.06 06:04:00.515740 [ 634143 ] {} <Fatal> BaseDaemon:
2020.07.06 06:04:00.515796 [ 634143 ] {} <Fatal> BaseDaemon: Stack trace: 0x7fc598ae13eb 0x7fc598ac0899 0x1eae83d2 0x22a95b18 0x22a95aa9 0x1564ea4b 0x1f066bdc 0x1f06691c 0x1f3df696 0x1f3e0122 0x1f97bae0 0x1f97aa5b 0x20280d8f 0x2029543a 0x2182800a 0x21828e8b 0
2020.07.06 06:04:00.515887 [ 634143 ] {} <Fatal> BaseDaemon: 3. raise @ 0x463eb in /usr/lib/x86_64-linux-gnu/libc-2.30.so
2020.07.06 06:04:00.515924 [ 634143 ] {} <Fatal> BaseDaemon: 4. abort @ 0x25899 in /usr/lib/x86_64-linux-gnu/libc-2.30.so
2020.07.06 06:04:00.517735 [ 634143 ] {} <Fatal> BaseDaemon: 5. /build/obj-x86_64-linux-gnu/../base/daemon/BaseDaemon.cpp:389: terminate_handler() @ 0x1eae83d2 in /usr/bin/clickhouse
2020.07.06 06:04:00.521493 [ 634143 ] {} <Fatal> BaseDaemon: 6. /build/obj-x86_64-linux-gnu/../contrib/libcxxabi/src/cxa_handlers.cpp:61: std::__terminate(void (*)()) @ 0x22a95b18 in /usr/bin/clickhouse
2020.07.06 06:04:00.525215 [ 634143 ] {} <Fatal> BaseDaemon: 7. /build/obj-x86_64-linux-gnu/../contrib/libcxxabi/src/include/atomic_support.h:78: std::terminate() @ 0x22a95aa9 in /usr/bin/clickhouse
2020.07.06 06:04:00.529245 [ 634143 ] {} <Fatal> BaseDaemon: 8. ? @ 0x1564ea4b in /usr/bin/clickhouse
2020.07.06 06:04:00.531095 [ 634143 ] {} <Fatal> BaseDaemon: 9. /build/obj-x86_64-linux-gnu/../base/common/../ext/scope_guard.h:47: ext::basic_scope_guard<DB::DatabaseWithDictionaries::createDictionary(DB::Context const&, std::__1::basic_string<char, std::__1
2020.07.06 06:04:00.532962 [ 634143 ] {} <Fatal> BaseDaemon: 10. /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/string:1426: DB::DatabaseWithDictionaries::createDictionary(DB::Context const&, std::__1::basic_string<char, std::__1::char_traits<char>, st
2020.07.06 06:04:00.535096 [ 634143 ] {} <Fatal> BaseDaemon: 11. /build/obj-x86_64-linux-gnu/../src/Interpreters/InterpreterCreateQuery.cpp:770: DB::InterpreterCreateQuery::createDictionary(DB::ASTCreateQuery&) @ 0x1f3df696 in /usr/bin/clickhouse
2020.07.06 06:04:00.537214 [ 634143 ] {} <Fatal> BaseDaemon: 12. /build/obj-x86_64-linux-gnu/../src/Interpreters/InterpreterCreateQuery.cpp:0: DB::InterpreterCreateQuery::execute() @ 0x1f3e0122 in /usr/bin/clickhouse
2020.07.06 06:04:00.539351 [ 634143 ] {} <Fatal> BaseDaemon: 13. /build/obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, DB::ReadBuffer*) @ 0x
2020.07.06 06:04:00.541522 [ 634143 ] {} <Fatal> BaseDaemon: 14. /build/obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:653: DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&,
```

**Comment by alexey-milovidov** [comment]
@tavplubix Don't understand why someone removes the `.sql.tmp` file.

**Comment by tavplubix** [comment]
Possible reasons:
- Exception from [`reloadDictionaryConfig(...)`](https://github.com/ClickHouse/ClickHouse/blob/99e9b15e00695cbb8532b56cf4fd57c271cddccf/src/Databases/DatabaseWithDictionaries.cpp#L181) after the file was successfully renamed
- Race with [`DROP DATABASE`](https://github.com/ClickHouse/ClickHouse/blob/948af0bfc51c3a56207b7cd4bfcee356b6216080/src/Databases/DatabaseOnDisk.cpp#L353)

Is this trace from the failed stress test check? I cannot grep any traces from the logs of failed check.

**Comment by alexey-milovidov** [comment]
> Is this trace from the failed stress test check?

Yes.

I download server log and grep for `###`.

**Comment by tavplubix** [comment]
Fixed in #13756
